File: 01-Programming-Languages/PHP/Data-Structures/Graph/graph_adjacency_list.php
<?php

/*
Problem: Graph Adjacency List
Description: Implement a graph with addEdge, printGraph, and BFS traversal using an adjacency list.

Approach:
- Use an associative array mapping vertices to lists of neighbors
- addEdge adds connection in both directions (undirected)
- printGraph displays all connections

Time Complexity: O(V + E) for traversal
Space Complexity: O(V + E)

Example:
Input:  vertices: [0,1,2,3], edges: [0-1, 0-2, 1-2, 2-3]
Output: adjacency list representation
*/

class Graph {
    private array $adjList;

    public function __construct() {
        $this->adjList = [];
    }

    public function addVertex($v): void {
        if (!isset($this->adjList[$v])) {
            $this->adjList[$v] = [];
        }
    }

    public function addEdge($v1, $v2): void {
        $this->addVertex($v1);
        $this->addVertex($v2);
        $this->adjList[$v1][] = $v2;
        $this->adjList[$v2][] = $v1;
    }

    public function printGraph(): void {
        foreach ($this->adjList as $v => $neighbors) {
            echo "$v -> " . implode(", ", $neighbors) . "\n";
        }
    }

    public function bfs($start): array {
        $visited = [];
        $queue = [$start];
        $visited[$start] = true;
        $result = [];
        while (!empty($queue)) {
            $node = array_shift($queue);
            $result[] = $node;
            foreach ($this->adjList[$node] as $neighbor) {
                if (!isset($visited[$neighbor])) {
                    $visited[$neighbor] = true;
                    $queue[] = $neighbor;
                }
            }
        }
        return $result;
    }
}

$g = new Graph();
$g->addEdge(0, 1);
$g->addEdge(0, 2);
$g->addEdge(1, 2);
$g->addEdge(2, 3);

echo "Adjacency List:\n";
$g->printGraph();

echo "BFS from 0: " . implode(" ", $g->bfs(0)) . "\n";

File: 01-Programming-Languages/Ruby/Data-Structures/Graph/graph_adjacency_list.rb
=begin
Problem: Graph Adjacency List
Description: Implement a graph with add_edge, print_graph, and bfs traversal using an adjacency list.

Approach:
- Use a hash mapping vertices to lists of neighbors
- add_edge adds connection in both directions

Time Complexity: O(V + E) for traversal
Space Complexity: O(V + E)

Example:
Input:  vertices: [0,1,2,3], edges: [0-1, 0-2, 1-2, 2-3]
Output: adjacency list representation
=end

class Graph
  def initialize
    @adj_list = {}
  end

  def add_vertex(v)
    @adj_list[v] ||= []
  end

  def add_edge(v1, v2)
    add_vertex(v1)
    add_vertex(v2)
    @adj_list[v1] << v2
    @adj_list[v2] << v1
  end

  def print_graph
    @adj_list.each { |v, neighbors| puts "#{v} -> #{neighbors.join(', ')}" }
  end

  def bfs(start)
    visited = Set.new
    queue = [start]
    visited.add(start)
    result = []
    until queue.empty?
      node = queue.shift
      result << node
      @adj_list[node].each do |neighbor|
        unless visited.include?(neighbor)
          visited.add(neighbor)
          queue << neighbor
        end
      end
    end
    result
  end
end

require 'set'

g = Graph.new
g.add_edge(0, 1)
g.add_edge(0, 2)
g.add_edge(1, 2)
g.add_edge(2, 3)

puts "Adjacency List:"
g.print_graph

puts "BFS from 0: #{g.bfs(0).join(' ')}"
